File: web-api/src/main/java/cn/labzen/web/api/response/Results.java
package cn.labzen.web.api.response;

import cn.labzen.web.api.definition.HttpStatusExt;
import org.springframework.http.HttpStatus;

public final class Results {

  private Results() {
  }

  public static Result success() {
    return new Result(200, null, null);
  }

  public static Result failure() {
    return new Result(500, null, null);
  }

  public static Result failure(String message) {
    return new Result(500, null, message);
  }

  /* ======================== with status ======================== */

  public static Result status(int code) {
    return new Result(code, null, null);
  }

  public static Result status(HttpStatus status) {
    return new Result(status.value(), null, null);
  }

  public static Result status(HttpStatusExt status) {
    return new Result(status.code(), null, null);
  }

  /* ======================== with value ======================== */

  public static Result with(Object value) {
    return new Result(200, value, null);
  }

  public static Result with(int code, Object value) {
    return new Result(code, value, null);
  }

  public static Result with(HttpStatus status, Object value) {
    return new Result(status.value(), value, null);
  }

  public static Result with(HttpStatusExt status, Object value) {
    return new Result(status.code(), value, null);
  }

  public static Result with(int code, Object value, String message) {
    return new Result(code, value, message);
  }

  public static Result with(HttpStatus status, Object value, String message) {
    return new Result(status.value(), value, message);
  }

  public static Result with(HttpStatusExt status, Object value, String message) {
    return new Result(status.code(), value, message);
  }

  /* ======================== with message ======================== */

  public static Result message(String message) {
    return new Result(200, null, message);
  }

  public static Result message(int code, String message) {
    return new Result(code, null, message);
  }

  public static Result message(HttpStatus status, String message) {
    return new Result(status.value(), null, message);
  }

  public static Result message(HttpStatusExt status, String message) {
    return new Result(status.code(), null, message);
  }

//  /* ======================== with paging records ======================== */
//
//  public static Result asPaging(List<?> records, int pageNumber, int pageSize) {
//    return asPaging(records, pageNumber, pageSize, null);
//  }
//
//  public static Result asPaging(List<?> records, int pageNumber, int pageSize, Long totalOfRecords) {
//    long totalOfPages = 0;
//    if (totalOfRecords != null) {
//      totalOfPages = totalOfRecords / pageSize + ((totalOfRecords % pageSize == 0) ? 0 : 1);
//    }
//
//    Pagination<?> pagination = new DefaultPagination<>(true, pageNumber, pageSize, totalOfRecords == null ? 0 : totalOfRecords, totalOfPages, records);
//    return new Result(200, pagination, null);
//  }
}

File: src/main/kotlin/cn/labzen/web/response/bean/Results.kt
package cn.labzen.web.response.bean

import cn.labzen.web.paging.Pagination
import cn.labzen.web.response.HttpStatusExt
import org.springframework.http.HttpStatus

object Results {

  @JvmStatic
  fun justSuccess() =
    Result(200)

  @JvmStatic
  fun justFailure() =
    Result(500)

  /* ======================== with status ======================== */

  @JvmStatic
  fun justStatus(status: Int) =
    Result(status)

  @JvmStatic
  fun justStatus(status: HttpStatus) =
    Result(status.value())

  @JvmStatic
  fun justStatus(status: HttpStatusExt) =
    Result(status.code)

  /* ======================== with value ======================== */

  @JvmStatic
  fun with(value: Any) =
    Result(200, value)

  @JvmStatic
  fun with(status: Int, value: Any) =
    Result(status, value)

  @JvmStatic
  fun with(status: HttpStatus, value: Any) =
    Result(status.value(), value)

  @JvmStatic
  fun with(status: HttpStatusExt, value: Any) =
    Result(status.code, value)

  @JvmStatic
  fun with(status: Int, value: Any, message: String) =
    Result(status, value, message)

  @JvmStatic
  fun with(status: HttpStatus, value: Any, message: String) =
    Result(status.value(), value, message)

  @JvmStatic
  fun with(status: HttpStatusExt, value: Any, message: String) =
    Result(status.code, value, message)

  /* ======================== with message ======================== */

  @JvmStatic
  fun withMessage(message: String) =
    Result(200, message = message)

  @JvmStatic
  fun withMessage(status: Int, message: String) =
    Result(status, message = message)

  @JvmStatic
  fun withMessage(status: HttpStatus, message: String) =
    Result(status.value(), message = message)

  @JvmStatic
  fun withMessage(status: HttpStatusExt, message: String) =
    Result(status.code, message = message)

  /* ======================== with paging records ======================== */

  @JvmStatic
  fun asPaging(records: List<*>, pageNumber: Int, pageSize: Int, totalOfRecords: Long? = null): Result {
    val totalOfPages = totalOfRecords?.let {
      it / pageSize + if (it % pageSize == 0L) 0 else 1
    } ?: 0
    val pagination = Pagination(pageNumber, pageSize, totalOfRecords ?: 0, totalOfPages, records)
    return Result(200, pagination)
  }
}